pkg/osquery/extension.go

```diff
@@ -304,6 +305,124 @@ func isNodeInvalidErr(err error) bool {
 	}
 }
 
+func (e *Extension) addRegistration(nodeKey string, enrollmentSecret string) {
+	// Extract munemo from enrollment secret.
+	// We do not have the key, and thus cannot verify -- so we use ParseUnverified.
+	token, _, err := new(jwt.Parser).ParseUnverified(enrollmentSecret, jwt.MapClaims{})
+	if err != nil {
+		e.slogger.Log(context.TODO(), slog.LevelWarn,
+			"could not parse enrollment secret",
+			"err", err,
+		)
+		return
+	}
+	claims, ok := token.Claims.(jwt.MapClaims)
+	if !ok {
+		e.slogger.Log(context.TODO(), slog.LevelWarn,
+			"no claims in enrollment secret",
+		)
+		return
+	}
+	munemo, munemoFound := claims["organization"]
+	if !munemoFound {
+		e.slogger.Log(context.TODO(), slog.LevelWarn,
+			"no claim for organization in enrollment secret",
+		)
+		return
+	}
+
+	r := types.Registration{
+		RegistrationID:   e.registrationId,
+		Munemo:           fmt.Sprintf("%s", munemo),
+		NodeKey:          nodeKey,
+		EnrollmentSecret: enrollmentSecret,
+	}
+
+	rawRegistration, err := json.Marshal(r)
+	if err != nil {
+		e.slogger.Log(context.TODO(), slog.LevelWarn,
+			"could not marshal registration",
+			"err", err,
+		)
+	}
+
+	if err := e.knapsack.RegistrationStore().Set([]byte(e.registrationId), rawRegistration); err != nil {
+		e.slogger.Log(context.TODO(), slog.LevelWarn,
+			"could not add registration",
+			"err", err,
+		)
+	}
+
+	e.slogger.Log(context.TODO(), slog.LevelInfo,
+		"successfully stored new registration",
+		"munemo", r.Munemo,
+	)
+}
+
+func (e *Extension) updateRegistration(nodeKey string) {
```

Comment:
just a few non-blocking thoughts/comments if any seem worth it to you- I don't have as strong of a vision of what this will look like next so might be off here:
- should `updateRegistration`/`addRegistration` at least have an error return? It doesn't seem necessary (i.e. an error shouldn't change any of the caller flows) currently so I could go either way, but this seemed like it could be an important enough path down the road to check
- it took me a little to read through and understand the update vs add Registration methods and grasp what flows should trigger which and when. would it at least sound clearer to make `updateRegistration` something like `ensureRegistrationNodeKeySet`? maybe with a func comment that explains it reads enroll secret and triggers `addRegistration` if needed

Reply:
Both add and update are under-commented, I will fix that! I could go either way on the error return too -- I think it sounds like a good idea for the future, so I will update.

Reply:
Added comments, renamed function!